### PR TITLE
Prevent ODE when ReadStart/Stop() is queued after disposal

### LINKED ISSUE #1327: Crash under high load
We have a *reproducible* crash on Kestrel when running a load test using Kestrel 1.1.0.

     ./wrk -d10s -t 32 -c 512 http://127.0.0.1:8080 -s batch-writes.lua -- 128 128

This generates 128 pipelined requests or about 30 KB each, on a 10 GiB network.

```
Unhandled Exception: System.ObjectDisposedException: Safe handle has been closed
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.Libuv.NativeMethods.uv_read_stop(UvStreamHandle handle)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.Libuv.read_stop(UvStreamHandle handle)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.UvStreamHandle.ReadStop()
   at Microsoft.AspNetCore.Server.Kestrel.Internal.KestrelThread.DoPostWork()
   at Microsoft.AspNetCore.Server.Kestrel.Internal.KestrelThread.OnPost()
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.Libuv.NativeMethods.uv_run(UvLoopHandle handle, Int32 mode)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.Libuv.run(UvLoopHandle handle, Int32 mode)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.KestrelThread.ThreadStart(Object parameter)
```

### PR DESCRIPTION
**Pull request by halter73**
Fixes #1327.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs [hunk 1]
```diff
@@ -285,23 +285,50 @@ private void OnRead(UvStreamHandle handle, int status)
...
+                {
+                    try
+                    {
+                        socket.ReadStart(_allocCallback, _readCallback, this);
+                    }
+                    catch (UvException)
+                    {
+                        // ReadStart() can throw a UvException in some cases (e.g. socket is no longer connected).
+                        // This should be treated the same as OnRead() seeing a "normalDone" condition.
+                        connection.Log.ConnectionReadFin(ConnectionId);
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: connection.ConnectionId

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs [hunk 2]
```diff
@@ -285,23 +285,50 @@ private void OnRead(UvStreamHandle handle, int status)
...
-            }
+                var connection = (Connection)thisObj;
+                var socket = connection._socket;
+
+                // It's possible that uv_close was called even before the call to Resume().
+                if (!socket.IsClosed)
+                {
+                    try
+                    {
+                        socket.ReadStart(_allocCallback, _readCallback, this);
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: this -> thisObj

**Comment by cesarbs** [comment]
Filed https://github.com/aspnet/KestrelHttpServer/issues/1340 for the AppVeyor failure.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/BufferSizeControl.cs [hunk 3]
```diff
@@ -1,4 +1,5 @@
-﻿using System.Diagnostics;
+﻿using System;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/BufferSizeControl.cs]: Unrelated but should be here: add copyright header.

File: test/Microsoft.AspNetCore.Server.KestrelTests/KestrelServerTests.cs [hunk 4]
```diff
@@ -23,7 +23,7 @@ public class KestrelServerTests
         [InlineData(-1337)]
         public void StartWithNonPositiveThreadCountThrows(int threadCount)
         {
-            var testLogger = new TestApplicationErrorLogger();
+            var testLogger = new TestApplicationErrorLogger { PrintCriticalErrors = false };
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KestrelServerTests.cs]: Why do you disable it only on these tests?

Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KestrelServerTests.cs]: I want to see critical errors in the test output only when they are unexpected. Arguably we should actually change this to ThrowOnCriticalErrors so we notice a test failure. Especially now that the console test output is much noisier by default.

**Comment by cesarbs** [comment]
AppVeyor failed again. @halter73 have you looked into that? I thought it was unrelated the first time it failed, but now it's looking suspicious.

**Comment by halter73** [comment]
@cesarbs I just reran the tests on AppVeyor without incident. I really don't think the errors could be related to this change. The main only difference aside from logging and test changes, is that we now check the uv_stream_t is still open before calling read_start/stop in certain cases.

The original failed test logs are [here](https://ci.appveyor.com/project/aspnetci/kestrelhttpserver/build/1.0.4095) for reference.

**Comment by cesarbs** [comment]
@halter73 I've added the link to https://github.com/aspnet/KestrelHttpServer/issues/1340. So far I haven't been able to reproduce any of those failures :confused:

**Pull request merged**